lld/MachO/ObjC.cpp

```diff
@@ -186,13 +186,24 @@ ObjcCategoryChecker::ObjcCategoryChecker()
       roClassLayout(target->wordSize), listHeaderLayout(target->wordSize),
       methodLayout(target->wordSize) {}
 
-// \p r must point to an offset within a cstring section.
+// \p r must point to an offset within a CStringInputSection or a
+// ConcatInputSection
 static StringRef getReferentString(const Reloc &r) {
   if (auto *isec = r.referent.dyn_cast<InputSection *>())
     return cast<CStringInputSection>(isec)->getStringRefAtOffset(r.addend);
   auto *sym = cast<Defined>(r.referent.get<Symbol *>());
-  return cast<CStringInputSection>(sym->isec())
-      ->getStringRefAtOffset(sym->value + r.addend);
+  if (isa<CStringInputSection>(sym->isec()))
+    return cast<CStringInputSection>(sym->isec())
+        ->getStringRefAtOffset(sym->value + r.addend);
+
+  if (isa<ConcatInputSection>(sym->isec())) {
+    auto &data = sym->isec()->data;
+    const char *pDataStart = reinterpret_cast<const char *>(data.data());
+    uint32_t len = strnlen(pDataStart + sym->value, data.size() - sym->value);
+    return StringRef(pDataStart + sym->value, len);
```

Comment:
I think this might be equivalent.

```suggestion
    auto &data = sym->isec()->data.slice(sym->value);
    retur StringRef(data.data(), data.size());
```

Reply:
The `concatInputSection` can have multiple strings in it - so can't use `data.slice(sym->value);`. Need to do a `strlen` first to determine the string size. I am not sure if `sim->size` is valid and can be used, but I can check if that is preferable.

Reply:
Ah I see. You can still use `data.slice(sym->value)` and then call `strnlen()`. It's much easier to work with `ArrayRef` then using raw `char *`.

Reply:
How about latest update ? Can't use `sym->size` since that will contain 0-padding for alignment.